### ISSUE TITLE
convolution bug

### ISSUE THREAD

**Issue by online-translation** [tensorlayer/tensorlayer#797]
I am using the latest tensorlayer in linux system.

There are two bugs that I want to report:

(1) binary conv2d can not work! 

(2) deform convolution can be used to train, but when evaluate, weights of offset parts were not included, which makes it impossible to load weight to evaluate.

**Comment by DEKHTIARJonathan** [comment]
Wow... This seems like a very huge bug.

Could you please give us a short snippet to test both cases. We will release a bug fix as soon as possible.

**Comment by online-translation** [comment]
(1) Binary
Some code like this:
```python
import tensorlayer as tl
import tensorflow as tf
from tensorlayer.layers import *

t_image = tf.placeholder(tf.float32, [None,128,128,1])
h_image = tf.placeholder(tf.float32, [None,128,128,1])**
net = InputLayer(t_image)
net = BinaryConv2d(net)
sess = tf.InteractiveSession()
train_op = tf.train.AdamOptimizer(learning_rate=LR).minimize(cost, var_list=model.all_params)
tl.layers.initialize_global_variables(sess)
cost = tl.cost.mean_squared_error(net.outputs, h_image)
for i in range(Max_iter):
    tr, hr = dataloader.get_next_train_batch()
    cost = sess.run([train_op, net.outputs], feed_dict = {t_image:tr, h_image:hr})
```
This will give you some error like: tensor can not be gradient.
(2) for the second issue, I try to add code at line 148 in deformable_conv.py. And it works:
```python
        self._add_layers(self.outputs)
        new_variables = get_collection_trainable(offset_layer.name)
        self._add_params(new_variables)
        
        if b_init:
            self._add_params([W, b])
        else:
            self._add_params(W)
```

**Comment by lgarithm** [comment]
For the case 1), could you give the complete example code?
Some variables are not defined in the above sample code.

**Comment by online-translation** [comment]
Sorry for that. @lgarithm. Following is the test code. 
```python
import tensorlayer as tl
import tensorflow as tf
from tensorlayer.layers import *
import numpy as np
t_image = tf.placeholder(tf.float32, [None,128,128,1])
h_image = tf.placeholder(tf.float32, [None,128,128,1])
net = InputLayer(t_image)
net = BinaryConv2d(net)
sess = tf.InteractiveSession()
cost = tl.cost.mean_squared_error(net.outputs, h_image)
train_op = tf.train.AdamOptimizer(learning_rate=1e-3).minimize(cost, var_list=net.all_params)
tl.layers.initialize_global_variables(sess)

for i in range(1):
    tr = np.random.randn(32,128,128,1)
    hr = np.random.randn(32,128,128,1)
    cost = sess.run([train_op, net.outputs], feed_dict = {t_image:tr, h_image:hr})
```
These codes will throw error like:

> NotImplementedError                       Traceback (most recent call last)
> <ipython-input-1-69743098cf3e> in <module>()
>       9 sess = tf.InteractiveSession()
>      10 cost = tl.cost.mean_squared_error(net.outputs, h_image)
> ---> 11 train_op = tf.train.AdamOptimizer(learning_rate=1e-3).minimize(cost, var_list=net.all_params)
>      12 tl.layers.initialize_global_variables(sess)
>      13 
> 
> ~/anaconda3/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py in minimize(self, loss, global_step, var_list, gate_gradients, aggregation_method, colocate_gradients_with_ops, name, grad_loss)
>     408 
>     409     return self.apply_gradients(grads_and_vars, global_step=global_step,
> --> 410                                 name=name)
>     411 
>     412   def compute_gradients(self, loss, var_list=None,
> 
> ~/anaconda3/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py in apply_gradients(self, grads_and_vars, global_step, name)
>     603           scope_name = var.op.name
>     604         with ops.name_scope("update_" + scope_name), ops.colocate_with(var):
> --> 605           update_ops.append(processor.update_op(self, grad))
>     606       if global_step is None:
>     607         apply_updates = self._finish(update_ops, name)
> 
> ~/anaconda3/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py in update_op(self, optimizer, g)
>     187 
>     188   def update_op(self, optimizer, g):
> --> 189     raise NotImplementedError("Trying to update a Tensor ", self._v)
>     190 
>     191 
> 
> NotImplementedError: ('Trying to update a Tensor ', <tf.Tensor 'binary_cnn2d/Sign:0' shape=(3, 3, 1, 32) dtype=float32>)

**Comment by lgarithm** [comment]
BinaryConv2d can't do backward, @XJTUWYD is this by design?

**Comment by DEKHTIARJonathan** [comment]
This is due to quantize op

https://github.com/tensorlayer/tensorlayer/blob/69f3f3693b0d7e35f1910fbea3c1ac9c97583cc2/tensorlayer/layers/utils.py#L345

```python
def quantize(x):
    # ref: https://github.com/AngusG/tensorflow-xnor-bnn/blob/master/models/binary_net.py#L70
    #  https://github.com/itayhubara/BinaryNet.tf/blob/master/nnUtils.py
    with tf.get_default_graph().gradient_override_map({"Sign": "TL_Sign_QuantizeGrad"}):
        return tf.sign(x)
```

**Comment by XJTUWYD** [comment]
The link of binary connect network is:[https://arxiv.org/pdf/1511.00363.pdf](url)
In this paper, author only use binary weight in forward and backward propagation but when it comes to update author update the gradient to original weights.
`def binarize(x):
    """
    Clip and binarize tensor using the straight through estimator (STE) for the gradient.
    """
    g = tf.get_default_graph()

    with ops.name_scope("Binarized") as name:
        with g.gradient_override_map({"Sign": "Identity"}):
            x=tf.clip_by_value(x,-1,1)
            return tf.sign(x)`

**Comment by zsdonghao** [comment]
fixed by 2.0.0 version
